solr/contrib/ltr/src/java/org/apache/solr/ltr/interleaving/TeamDraftInterleaving.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.ltr.interleaving;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Random;
+import java.util.Set;
+
+import org.apache.lucene.search.ScoreDoc;
+
+public class TeamDraftInterleaving implements Interleaving{
+  public static Random RANDOM;
+
+  static {
+    // We try to make things reproducible in the context of our tests by initializing the random instance
+    // based on the current seed
+    String seed = System.getProperty("tests.seed");
```

Comment:
11.2/n `LTRQParserPlugin.LTRQParser` also has access to the `SolrQueryRequest` and its `SolrCore` object. For some reason I thought that within that some 'official' source of random-ness might be available which could be passed to a `TeamDraftInterleaving(Random)` constructor. And I imagined that our test harnesses would use seeds to make tests reproducible w.r.t. that 'official' source of random-ness. There however doesn't appear to be such a source of non-test official random-ness? `System.getProperty("tests.seed");` being used/available to non-test code seems potentially tricky.

@dweiss would you perhaps have any insights around non-test sources of randomness?

Reply:
I spent quite a lot of time on this, to make it compatible with the tests, if anyone else has some better solution I would be more than happy to change it, I was not super happy of what I ended with, but it was the only solution I found to be working at the time.

Reply:
Yes, predictable randomness in tests can be tricky! I had an (unplanned) idea around this at breakfast this morning, trying it out now: https://github.com/cpoerschke/lucene-solr/commit/5fc99898f9bc82f4dabcfe188e132ba2bc727704 is part 1 of 2 (and is independent technically speaking).

Reply:
https://github.com/cpoerschke/lucene-solr/commit/63a190b57fdc892b98c546f005edbc096e4e4095 is part 2 of 2: it removes the tests.seed System property use in TeamDraftInterleaving by passing a Random object into TeamDraftInterleaving. instead the tests.seed System property use is in a test class (LTRQParserTestPlugin) and tests continue to make their `setRANDOM` calls but instead of `TeamDraftInterleaving.setRANDOM` it's now `LTRQParserTestPlugin.setRANDOM` then.

Two problems with this though:
* a forbidden apis check fails (saying that RandomizedRunner's random() should be used instead but i'm unclear still on how that might be done)
* the tests no longer pass, which baffles me though haven't looked at details yet; speculations:
  * was the randomness in the tests predictable before and now it's predictable also but the number sequence has changed and test expectations need to be adjusted to match?
  * something to do with order of system property setting and non-test/test class loading perhaps and perhaps use of the RandomizedRunner's random would solve it somehow?

Reply:
Now I remember, I worked on this a lot and I ended up with the current solution even if including the test seed get property.
I have not introduced that approach on my own, but I took inspiration from other areas of Solr code:
such as:
org/apache/solr/cloud/api/collections/OverseerCollectionMessageHandler.java:174
org/apache/solr/core/BlobRepository.java:73
org/apache/solr/servlet/HttpSolrCall.java:152
org/apache/solr/util/tracing/GlobalTracer.java:35

So it is already used quite widely across Solr production code.
Unless you have stronger concerns we can probably leave it as it is and resolve the conversation here :)

Reply:
Ah, precedent and existing use already, good to know, thanks for sharing!